pvpanic/pvpanic/power.c

```diff
@@ -81,23 +84,65 @@ NTSTATUS PVPanicEvtDevicePrepareHardware(IN WDFDEVICE Device,
                     context->IoBaseAddress =
                         (PVOID)(ULONG_PTR)desc->u.Port.Start.QuadPart;
                 }
-                PvPanicPortAddress = (PUCHAR)context->IoBaseAddress;
+                if(!PvPanicPortOrMemAddress)
+                    PvPanicPortOrMemAddress = (PUCHAR)context->IoBaseAddress;
+                else
+                    return STATUS_DEVICE_CONFIGURATION_ERROR;
 
                 break;
             }
 
+            case CmResourceTypeMemory:
+            {
+                TraceEvents(TRACE_LEVEL_VERBOSE, DBG_POWER,
+                    "Memory mapped CSR: (%x) Length: (%d)",
+                    desc->u.Memory.Start.LowPart, desc->u.Memory.Length);
+
+                context->MemRange = desc->u.Memory.Length;
+#if defined(NTDDI_WINTHRESHOLD) && (NTDDI_VERSION >= NTDDI_WINTHRESHOLD)
+                context->MemBaseAddress = MmMapIoSpaceEx(
+                    desc->u.Memory.Start,
+                    desc->u.Memory.Length,
+                    PAGE_READWRITE | PAGE_NOCACHE);
+#else
+                context->MemBaseAddress = MmMapIoSpace(
+                    desc->u.Memory.Start,
+                    desc->u.Memory.Length,
+                    MmNonCached);
+#endif
+                if (!PvPanicPortOrMemAddress)
+                    PvPanicPortOrMemAddress = (PUCHAR)context->MemBaseAddress;
+                else
+                    return STATUS_DEVICE_CONFIGURATION_ERROR;
```

Comment:
I think some trace message can be printed to detect situation when 2nd PVPanic device is being added over existing one. It would be helpful when user accidentally added PCI and ISA devices together.

Reply:
Will add trace there, thanks.